### DISCUSSION TITLE
Standard comment format not respected in Golang

### DISCUSSION THREAD

**Discussion by junedev** [community/community#11494] [Copilot Conversations]
Comments for functions in Go are supposed to always start with the name of the function. However, Copilot is not respecting that rule. It will always suggest some other name than the actual function name as first word of the comment when the comment is written after the function.

**Comment by donno2048** [comment]
I suspect this is because the type of comments you're referring to is **docstrings** not **comments**, read more about godoc [here](https://pkg.go.dev/golang.org/x/tools/cmd/godoc), not every comment have to be a docstring...

**Reply by junedev** [reply]
Yes, exactly, I am referring to those docstrings. When I have an exported function, standard linters will complain when there is no comment/docstring in the correct format. Yet Copilot does not make a suggestion in that format.

**Reply by donno2048** [reply]
OK, but docstrings are just a formality there is no reason for copilot not to use standard comments...

**Comment by junedev** [comment]
@donno2048 I am not sure I follow you comment
> OK, but docstrings are just a formality there is no reason for copilot not to use standard comments...

To give some more background...

* Go does not have a different syntax for what you call "docstrings" and normal comments.
* Comments for exported functions in that specific format are a very integral part of writing good Go code. As mentioned, standard linters will complain.
* The rule I mentioned is part of the official language documentation (https://go.dev/doc/effective_go#commentary):
  > Doc comments work best as complete sentences, which allow a wide variety of automated presentations. The first sentence should be a one-sentence summary that starts with the name being declared
  >    ```
  >    // Compile parses a regular expression and returns, if successful,
  >    // a Regexp that can be used to match against text.
  >    func Compile(str string) (*Regexp, error) {
  >    ```

Instead of following this rule, here an example of what Copilot suggests:
![Screenshot from 2022-02-14 17-28-27](https://user-images.githubusercontent.com/12543047/153905996-771190cd-ee4b-4d9e-a12a-2e6685d6c226.png)

So Copilot does not follow the official rules for writing good Go code. I thought this is something the Copilot team would like to know.

**Reply by donno2048** [reply]
I see, that wasn't obvious from your original question, I thought you meant it doesn't use **only** docstrings in functions, my bad...

**Reply by davecheney** [reply]
I’m pretty sure if you type a few more letters copilot will give a better suggestion. Remember, artificial intelligence isn’t actually intelligence, it’s just a talented mimic.

**Reply by junedev** [reply]
@davecheney I am aware of that, however I would have assumed copilot could be able to pick up this Go pattern about the comments for exported functions and mimic that. This is why I reported the feedback here.

Is it your experience that it works well when filling in the comments? I am consistently seeing these wrong suggestions. E.g. I have a function `SetFieldInRequestObject`, if I type `// SetFieldIn` or even `// SetFieldInRequest` the suggestion does not show me the full function name.

Here is a screenshot of the example above when I type out the first word. (My cursor is after the `e` of `Calculate`.)
![Screenshot from 2022-02-14 20-44-18](https://user-images.githubusercontent.com/12543047/153935861-ca15722d-c262-4d1f-8313-40819482a32a.png)

**Reply by davecheney** [reply]
I find that if I wrote the comment first, copilot is pretty good at figuring out the function signature.

**Reply by junedev** [reply]
Yes, I agree. My feedback of course only applies for cases where someone forgot to write the comment first like it happens to me ever so often. :see_no_evil:

**Reply by KTibow** [reply]
I'm pretty sure the reason for this happening is that Copilot doesn't send the text that goes after your current line. You can check the prompt Copilot sends to the server in Terminal > Output > GitHub Copilot.

**Reply by junedev** [reply]
@KTibow Took me a while to get around to check this. Now I did and I wasn't able to see anything helpful in those logs. This is what I see there over and over again.
```
[INFO] [ghostText] [2022-02-24T16:36:35.667Z] Found inline suggestions locally
[INFO] [ghostText] [2022-02-24T16:36:35.815Z] Found inline suggestions locally
[INFO] [default] [2022-02-24T16:36:44.888Z] [fetchCompletions] engine https://copilot-proxy.githubusercontent.com/v1/engines/copilot-codex
[INFO] [default] [2022-02-24T16:36:45.349Z] request.response: [https://copilot-proxy.githubusercontent.com/v1/engines/copilot-codex/completions] took 460 ms
[INFO] [streamChoices] [2022-02-24T16:36:45.351Z] solution 1 returned. finish reason: ["stop"] finishOffset: [undefined] completionId: [{cmpl-4fBOeRUojHigoFM2LpEUAb30qnnmO}] created: [{1645720604}]
[INFO] [streamChoices] [2022-02-24T16:36:45.353Z] solution 0 returned. finish reason: ["stop"] finishOffset: [undefined] completionId: [{cmpl-4fBOeRUojHigoFM2LpEUAb30qnnmO}] created: [{1645720604}]
```

**Reply by KTibow** [reply]
If you press Ctrl+Enter to trigger the seperate suggestion pane, you should see it.

**Reply by junedev** [reply]
@KTibow Found some time to test this again.

I tried this code and opened the pane with the cursor after `Cal` in the comment.
```
package main

// Cal
func CalculateSum(numbers []int) int {
	sum := 0
	for _, number := range numbers {
		sum += number
	}
	return sum
}
```

In the suggestion pane I see the following but I don't understand how that gives an indication whether or not the code below the comment was considered when creating the suggestions (referring to your theory in https://github.com/github/feedback/discussions/11494#discussioncomment-2192662):

```
Synthesizing 5/10 solutions (Duplicates hidden)

=======

// Call the function main() to start the program.
func main() {
	// Println outputs a line to stdout.
	// Println formats using the default formats for its operands and
	// writes to standard output. Spaces are added between operands when
	// neither is a string. It returns the number of bytes written and any
	// write error encountered.
	print("Hello World!")
}

=======

// Call the function main() to start the program.
func main() {
	// Println outputs a line to stdout.
	// Arguments can be any number of values of any type.
	// Arguments are separated by spaces.
	// Println outputs a newline character.
	print("Hello World!")
}

=======

// Call the function main() to start the program.
func main() {
	// Println outputs a line to stdout.
	// Arguments can be given in any order.
	println("Hello World!")
}

=======

// Call the function main() to start the program.
func main() {
	// Println outputs a line to stdout.
	// Arguments can be any number of values of any type.
	println("Hello, World!")
}
```

**Reply by KTibow** [reply]
> In the suggestion pane I see the following but I don't understand how that gives an indication whether or not the code below the comment was considered when creating the suggestions (referring to your theory in https://github.com/github/feedback/discussions/11494#discussioncomment-2192662):

@junedev I was talking about the output pane, in Terminal > Output > GitHub Copilot.

**Reply by junedev** [reply]
@KTibow Sorry, I didn't expect the output to change based on whether the suggestion pane is open or not.

I think I found the log entry you are referring to:
```
[INFO] [solutions] [2022-04-14T06:00:47.108Z] prompt: {"prefix":"// Language: go\n// Path: main.go\npackage main\n\n// Cal","suffix":""}
```

I checked other cases in some real code as well and suffix is always empty. So is that a bug?

Full output below.
```
[INFO] [default] [2022-04-14T06:00:44.966Z] [fetchCompletions] engine https://copilot-proxy.githubusercontent.com/v1/engines/copilot-codex
[INFO] [default] [2022-04-14T06:00:45.165Z] request.response: [https://copilot-proxy.githubusercontent.com/v1/engines/copilot-codex/completions] took 199 ms
[INFO] [streamChoices] [2022-04-14T06:00:45.167Z] solution 0 returned. finish reason: ["stop"] finishOffset: [undefined] completionId: [{cmpl-4wmp36AuxyOMgba09Lw43u0T11dKj}] created: [{1649916045}]
[INFO] [streamChoices] [2022-04-14T06:00:45.169Z] solution 1 returned. finish reason: ["stop"] finishOffset: [undefined] completionId: [{cmpl-4wmp36AuxyOMgba09Lw43u0T11dKj}] created: [{1649916045}]
[INFO] [ghostText] [2022-04-14T06:00:45.721Z] Found inline suggestions locally
[INFO] [solutions] [2022-04-14T06:00:47.108Z] prompt: {"prefix":"// Language: go\n// Path: main.go\npackage main\n\n// Cal","suffix":""}
[INFO] [default] [2022-04-14T06:00:47.110Z] [fetchCompletions] engine https://copilot-proxy.githubusercontent.com/v1/engines/copilot-codex
[INFO] [default] [2022-04-14T06:00:47.378Z] request.response: [https://copilot-proxy.githubusercontent.com/v1/engines/copilot-codex/completions] took 268 ms
[INFO] [streamChoices] [2022-04-14T06:00:52.283Z] solution 1 returned. finish reason: ["length"] finishOffset: [undefined] completionId: [{cmpl-4wmp5l4wIabdWhIUE94CPWwdot5AX}] created: [{1649916047}]
[INFO] [solutions] [2022-04-14T06:00:52.290Z] Filtered out repetitive solution
[INFO] [streamChoices] [2022-04-14T06:00:52.291Z] solution 7 returned. finish reason: ["stop"] finishOffset: [undefined] completionId: [{cmpl-4wmp5l4wIabdWhIUE94CPWwdot5AX}] created: [{1649916047}]
[INFO] [solutions] [2022-04-14T06:00:52.292Z] Open Copilot completion: [l the function main() to start the program.
func main() {
	// Println outputs a line to stdout.
	// Arguments can be given in any order.
	println("Hello World!")
}]
[INFO] [streamChoices] [2022-04-14T06:00:52.293Z] solution 5 returned. finish reason: ["length"] finishOffset: [undefined] completionId: [{cmpl-4wmp5l4wIabdWhIUE94CPWwdot5AX}] created: [{1649916047}]
[INFO] [solutions] [2022-04-14T06:00:52.296Z] Filtered out repetitive solution
[INFO] [streamChoices] [2022-04-14T06:00:52.297Z] solution 6 returned. finish reason: ["stop"] finishOffset: [undefined] completionId: [{cmpl-4wmp5l4wIabdWhIUE94CPWwdot5AX}] created: [{1649916047}]
[INFO] [solutions] [2022-04-14T06:00:52.297Z] Open Copilot completion: [l the function main() to start the program.
func main() {
	// Println outputs a line to stdout.
	// Println formats using the default formats for its operands and
	// writes to standard output. Spaces are added between operands when
	// neither is a string. It returns the number of bytes written and any
	// write error encountered.
	print("Hello World!")
}]
[INFO] [streamChoices] [2022-04-14T06:00:52.299Z] solution 8 returned. finish reason: ["length"] finishOffset: [undefined] completionId: [{cmpl-4wmp5l4wIabdWhIUE94CPWwdot5AX}] created: [{1649916047}]
[INFO] [solutions] [2022-04-14T06:00:52.303Z] Filtered out repetitive solution
[INFO] [streamChoices] [2022-04-14T06:00:52.303Z] solution 3 returned. finish reason: ["stop"] finishOffset: [undefined] completionId: [{cmpl-4wmp5l4wIabdWhIUE94CPWwdot5AX}] created: [{1649916047}]
[INFO] [solutions] [2022-04-14T06:00:52.304Z] Open Copilot completion: [l the function main() to start the program.
func main() {
	// Println outputs a line to stdout.
	// Println formats using the default formats for its operands and
	// writes to standard output. Spaces are added between operands when
	// neither is a string. It returns the number of bytes written and any
	// write error encountered.
	print("Hello World!")
}]
[INFO] [streamChoices] [2022-04-14T06:00:52.305Z] solution 2 returned. finish reason: ["stop"] finishOffset: [undefined] completionId: [{cmpl-4wmp5l4wIabdWhIUE94CPWwdot5AX}] created: [{1649916047}]
[INFO] [solutions] [2022-04-14T06:00:52.305Z] Open Copilot completion: [l the function main() to start the program.
func main() {
	// Println outputs a line to stdout.
	// Println formats using the default formats for its operands and
	// writes to standard output. Spaces are always added between
	// operands and a newline is appended. It returns the number of bytes
	// written and any write error encountered.
	println("Hello, World!")
}]
[INFO] [streamChoices] [2022-04-14T06:00:52.307Z] solution 9 returned. finish reason: ["length"] finishOffset: [undefined] completionId: [{cmpl-4wmp5l4wIabdWhIUE94CPWwdot5AX}] created: [{1649916047}]
[INFO] [solutions] [2022-04-14T06:00:52.309Z] Filtered out repetitive solution
[INFO] [streamChoices] [2022-04-14T06:00:52.310Z] solution 0 returned. finish reason: ["length"] finishOffset: [undefined] completionId: [{cmpl-4wmp5l4wIabdWhIUE94CPWwdot5AX}] created: [{1649916047}]
[INFO] [solutions] [2022-04-14T06:00:52.312Z] Filtered out repetitive solution
[INFO] [streamChoices] [2022-04-14T06:00:52.314Z] solution 4 returned. finish reason: ["length"] finishOffset: [undefined] completionId: [{cmpl-4wmp5l4wIabdWhIUE94CPWwdot5AX}] created: [{1649916047}]
[INFO] [solutions] [2022-04-14T06:00:52.316Z] Filtered out repetitive solution
```

**Reply by KTibow** [reply]
> So is that a bug?

No, it's just how Codex works, or at least right now. `suffix` might be filled if Copilot is making a suggestion partway through a line of code, but it doesn't seem to be happening for me right now. I do agree that it would be useful if Codex could consider code that comes after, though.
